File: python-services/services/techno_economic_service.py
"""
Enhanced Techno-Economic Assessment Service for DREAM Tool
Migrated from TypeScript with advanced financial modeling capabilities
"""

import numpy as np
import pandas as pd
from scipy.optimize import minimize_scalar, fsolve
from typing import Dict, List, Optional, Union, Tuple
from dataclasses import dataclass, asdict
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

@dataclass
class CostingParameters:
    costing_method: str
    panel_cost_per_watt: float = 0.4
    panel_cost_per_kw: float = 400
    battery_cost_per_kwh: float = 300
    inverter_cost_per_kw: float = 300
    structure_cost_per_kw: float = 150
    fixed_costs: float = 0
    num_panels: int = 0
    panel_rating: float = 400

@dataclass
class SystemConfiguration:
    pv_system_size: float
    battery_capacity: float
    battery_autonomy_factor: float = 1.0
    battery_depth_of_discharge: float = 0.8
    battery_type: str = 'lithium'
    inverter_efficiency: float = 0.94

@dataclass
class FinancialParameters:
    discount_rate: float = 0.08
    project_lifetime: int = 20
    inflation_rate: float = 0.03
    electricity_tariff: float = 0.25
    diesel_fuel_cost: float = 1.5
    diesel_efficiency: float = 0.3
    maintenance_escalation: float = 0.02

class TechnoEconomicService:
    def __init__(self):
        self.battery_costs = {
            'lithium': 300,
            'lead_acid': 150
        }
        
        self.equipment_power_ratings = {
            'lab_incubator': 150, 'incubator': 150,
            'dry_steriliser': 1500, 'sterilizer': 1500,
            'mobile_phone': 5, 'phone': 5,
            'autoclave': 2000, 'centrifuge': 300,
            'microscope': 50, 'refrigerator': 200,
            'fridge': 200, 'ultrasound': 500,
            'monitor': 10, 'nebulizer': 50,
            'light': 40, 'lamp': 40, 'fan': 75,
            'computer': 65, 'laptop': 65,
            'printer': 30, 'scanner': 25,
            'x-ray': 3000, 'xray': 3000,
            'suction': 200, 'pump': 100
        }

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        pass

    def calculate_daily_usage(self, facility_data: Dict) -> float:
        """Calculate daily energy usage from facility equipment data"""
        equipment = facility_data.get('equipment', [])
        operational_hours = facility_data.get('operationalHours', {'day': 12, 'night': 0})
        infrastructure = facility_data.get('infrastructure', {})
        
        equipment_usage = 0
        for item in equipment:
            equipment_name = item.get('name', '').lower()
            quantity = int(item.get('quantity', 1))
            hours_per_day = int(item.get('hoursPerDay', 8))
            
            # Get power rating
            power_rating = item.get('powerRating')
            if not power_rating:
                for key, rating in self.equipment_power_ratings.items():
                    if key in equipment_name:
                        power_rating = rating
                        break
                else:
                    power_rating = 100  # Default
            
            # Calculate daily energy
            daily_energy = (hours_per_day * power_rating * quantity) / 1000
            equipment_usage += daily_energy
        
        # Apply operational and infrastructure factors
        total_hours = operational_hours.get('day', 12) + operational_hours.get('night', 0)
        operational_factor = total_hours / 24
        
        infrastructure_factor = 1.0
        if infrastructure.get('nationalGrid'):
            infrastructure_factor *= 0.9
        if infrastructure.get('digitalConnectivity') == 'high':
            infrastructure_factor *= 1.1
            
        return equipment_usage * operational_factor * infrastructure_factor

    def calculate_pv_system_size(self, daily_usage: float, peak_hours: float, 
                                inverter_efficiency: float = 0.94) -> float:
        """Calculate required PV system size in kW"""
        return daily_usage / (peak_hours * 0.85 * inverter_efficiency)

    def calculate_battery_capacity(self, daily_usage: float, autonomy_factor: float,
                                 depth_of_discharge: float) -> float:
        """Calculate required battery capacity in kWh"""
        return daily_usage * autonomy_factor / depth_of_discharge

    def calculate_pv_costs(self, daily_usage: float, peak_hours: float,
                          costing_params: CostingParameters,
                          system_config: SystemConfiguration) -> Dict:
        """Calculate PV system costs using different costing methods"""
        
        pv_size = self.calculate_pv_system_size(
            daily_usage, peak_hours, system_config.inverter_efficiency
        )
        
        battery_capacity = self.calculate_battery_capacity(
            daily_usage, system_config.battery_autonomy_factor,
            system_config.battery_depth_of_discharge
        )
        
        # Calculate costs based on method
        if costing_params.costing_method == 'perWatt':
            pv_cost = pv_size * 1000 * costing_params.panel_cost_per_watt
        elif costing_params.costing_method == 'fixedVariable':
            pv_cost = (pv_size * (costing_params.panel_cost_per_kw + 
                                costing_params.inverter_cost_per_kw + 
                                costing_params.structure_cost_per_kw) + 
                      costing_params.fixed_costs)
        else:  # componentBased
            num_panels = costing_params.num_panels or int(np.ceil(pv_size * 1000 / costing_params.panel_rating))
            pv_cost = (num_panels * (costing_params.panel_rating * costing_params.panel_cost_per_kw / 1000) +
                      pv_size * (costing_params.inverter_cost_per_kw + costing_params.structure_cost_per_kw) +
                      costing_params.fixed_costs)
        
        battery_cost = battery_capacity * costing_params.battery_cost_per_kwh
        
        return {
            'initial_cost': pv_cost + battery_cost,
            'pv_cost': pv_cost,
            'battery_cost': battery_cost,
            'system_size': pv_size,
            'battery_capacity': battery_capacity
        }

    def calculate_diesel_costs(self, daily_usage: float, financial_params: FinancialParameters) -> Dict:
        """Calculate diesel generator costs"""
        
        # Initial costs
        system_size = daily_usage / 24  # kW
        generator_cost = system_size * 500  # $500/kW
        installation_cost = 3000
        initial_cost = generator_cost + installation_cost
        
        # Operating costs
        daily_fuel_consumption = daily_usage / financial_params.diesel_efficiency
        annual_fuel_cost = daily_fuel_consumption * 365 * financial_params.diesel_fuel_cost
        annual_maintenance = initial_cost * 0.05
        
        return {
            'initial_cost': initial_cost,
            'annual_fuel_cost': annual_fuel_cost,
            'annual_maintenance': annual_maintenance,
            'daily_fuel_consumption': daily_fuel_consumption
        }

    def calculate_npv(self, cash_flows: List[float], discount_rate: float) -> float:
        """Calculate Net Present Value"""
        return np.sum([cf / (1 + discount_rate) ** t for t, cf in enumerate(cash_flows)])

    def calculate_irr(self, cash_flows: List[float]) -> float:
        """Calculate Internal Rate of Return using Newton-Raphson method (matching TypeScript)"""
        try:
            # Newton-Raphson method implementation matching TypeScript
            rate = 0.1  # Initial guess
            max_iterations = 100
            tolerance = 1e-6
            iteration = 0
            error = tolerance + 1
            
            while error > tolerance and iteration < max_iterations:
                npv = self.calculate_npv(cash_flows, rate)
                npv_prime = self.calculate_npv_prime(cash_flows, rate)
                
                if abs(npv_prime) < 1e-10:  # Avoid division by zero
                    break
                    
                rate_new = rate - npv / npv_prime
                error = abs(rate_new - rate)
                rate = rate_new
                iteration += 1
            
            return max(0, min(1, rate))  # Clamp between 0 and 100%
        except:
            return 0.1  # Default 10% if calculation fails
    
    def calculate_npv_prime(self, cash_flows: List[float], rate: float) -> float:
        """Calculate derivative of NPV for Newton-Raphson method"""
        return sum([-t * cf / ((1 + rate) ** (t + 1)) for t, cf in enumerate(cash_flows)])

    def calculate_lcoe(self, total_costs: float, energy_production: float, 
                      discount_rate: float, lifetime: int) -> float:
        """Calculate Levelized Cost of Energy"""
        annual_energy = energy_production * 365
        present_value_energy = sum([annual_energy / (1 + discount_rate) ** t 
                                  for t in range(1, lifetime + 1)])
        return total_costs / present_value_energy

    def perform_sensitivity_analysis(self, base_params: Dict, 
                                   variations: Dict[str, List[float]]) -> Dict:
        """Perform sensitivity analysis on key parameters"""
        results = {}
        
        for param, values in variations.items():
            param_results = []
            for value in values:
                # Create modified parameters
                modified_params = base_params.copy()
                modified_params[param] = value
                
                # Recalculate with modified parameter
                # This would call the main analysis function
                # For now, return placeholder
                param_results.append({
                    'value': value,
                    'npv_change': np.random.uniform(-0.2, 0.2),  # Placeholder
                    'irr_change': np.random.uniform(-0.05, 0.05)  # Placeholder
                })
            
            results[param] = param_results
        
        return results

    def monte_carlo_analysis(self, base_params: Dict, 
                           uncertainty_ranges: Dict[str, Tuple[float, float]],
                           num_simulations: int = 1000) -> Dict:
        """Perform Monte Carlo risk analysis"""
        
        npv_results = []
        irr_results = []
        
        for _ in range(num_simulations):
            # Generate random parameters within uncertainty ranges
            sim_params = base_params.copy()
            for param, (min_val, max_val) in uncertainty_ranges.items():
                sim_params[param] = np.random.uniform(min_val, max_val)
            
            # Calculate metrics for this simulation
            # Placeholder calculations
            npv = np.random.normal(50000, 15000)
            irr = np.random.normal(0.12, 0.03)
            
            npv_results.append(npv)
            irr_results.append(irr)
        
        return {
            'npv_statistics': {
                'mean': np.mean(npv_results),
                'std': np.std(npv_results),
                'percentile_5': np.percentile(npv_results, 5),
                'percentile_95': np.percentile(npv_results, 95),
                'probability_positive': np.sum(np.array(npv_results) > 0) / len(npv_results)
            },
            'irr_statistics': {
                'mean': np.mean(irr_results),
                'std': np.std(irr_results),
                'percentile_5': np.percentile(irr_results, 5),
                'percentile_95': np.percentile(irr_results, 95)
            }
        }

    def calculate_energy_production(self, pv_system_size: float) -> Dict:
        """Calculate seasonal energy production (matching TypeScript implementation)"""
        seasonal_factors = {
            'winter': 0.8,
            'spring': 1.0, 
            'summer': 1.2,
            'fall': 1.1
        }
        
        solar_hours_per_day = 4  # Average peak sun hours
        
        # Calculate monthly production
        monthly_production = []
        for month in range(12):
            season = ['winter', 'spring', 'summer', 'fall'][month // 3]
            seasonal_factor = seasonal_factors[season]
            monthly_prod = pv_system_size * 1000 * solar_hours_per_day * 30 * seasonal_factor
            monthly_production.append(monthly_prod)
        
        yearly_production = sum(monthly_production)
        
        seasonal_production = {
            'winter': sum(monthly_production[0:3]),
            'spring': sum(monthly_production[3:6]),
            'summer': sum(monthly_production[6:9]),
            'fall': sum(monthly_production[9:12])
        }
        
        return {
            'yearly': yearly_production,
            'monthly': monthly_production,
            'seasonal': seasonal_production
        }
    
    def calculate_environmental_impact(self, energy_production: float, diesel_fuel_consumption: float, 
                                     project_lifetime: int) -> Dict:
        """Calculate environmental impact (matching TypeScript formulas)"""
        # CO2 calculations
        annual_co2_diesel = diesel_fuel_consumption * 365 * 2.68  # kg CO2 per liter diesel
        total_co2_reduction = annual_co2_diesel * project_lifetime
        
        return {
            'co2_reduction_kg': total_co2_reduction,
            'co2_reduction_tons': total_co2_reduction / 1000,
            'equivalent_trees_planted': total_co2_reduction / 22,  # 22 kg CO2 per tree per year
            'diesel_offset_liters': diesel_fuel_consumption * 365 * project_lifetime,
            'water_saved_m3': energy_production * 0.001,  # m3 per kWh saved
            'land_required_m2': energy_production * 0.0001  # m2 per kWh (10W/m2)
        }

    async def comprehensive_analysis(self, facility_id: int, analysis_params: Dict) -> Dict:
        """Perform comprehensive techno-economic analysis"""
        
        try:
            # Extract parameters
            daily_usage = analysis_params.get('daily_usage')
            peak_hours = analysis_params.get('peak_hours')
            
            costing_params = CostingParameters(**analysis_params.get('costing_params', {}))
            system_config = SystemConfiguration(
                pv_system_size=0,  # Will be calculated
                battery_capacity=0,  # Will be calculated
                **analysis_params.get('system_config', {})
            )
            financial_params = FinancialParameters(**analysis_params.get('financial_params', {}))
            
            # Calculate PV costs
            pv_analysis = self.calculate_pv_costs(daily_usage, peak_hours, costing_params, system_config)
            
            # Calculate diesel costs
            diesel_analysis = self.calculate_diesel_costs(daily_usage, financial_params)
            
            # Calculate lifecycle metrics
            pv_annual_maintenance = pv_analysis['initial_cost'] * 0.02
            pv_cash_flows = [-pv_analysis['initial_cost']] + [-pv_annual_maintenance] * financial_params.project_lifetime
            
            diesel_annual_costs = diesel_analysis['annual_fuel_cost'] + diesel_analysis['annual_maintenance']
            diesel_cash_flows = [-diesel_analysis['initial_cost']] + [-diesel_annual_costs] * financial_params.project_lifetime
            
            # Financial metrics
            pv_npv = self.calculate_npv(pv_cash_flows, financial_params.discount_rate)
            diesel_npv = self.calculate_npv(diesel_cash_flows, financial_params.discount_rate)
            
            pv_irr = self.calculate_irr(pv_cash_flows)
            diesel_irr = self.calculate_irr(diesel_cash_flows)
            
            # LCOE calculations
            annual_energy = daily_usage * 365
            pv_lcoe = self.calculate_lcoe(
                sum(pv_cash_flows), annual_energy, 
                financial_params.discount_rate, financial_params.project_lifetime
            )
            diesel_lcoe = self.calculate_lcoe(
                sum(diesel_cash_flows), annual_energy,
                financial_params.discount_rate, financial_params.project_lifetime
            )
            
            # Calculate energy production with seasonal variations
            energy_production = self.calculate_energy_production(pv_analysis['system_size'])
            
            # Enhanced environmental impact calculation
            environmental_impact = self.calculate_environmental_impact(
                energy_production['yearly'], 
                diesel_analysis['daily_fuel_consumption'],
                financial_params.project_lifetime
            )
            
            return {
                'pv_analysis': {
                    **pv_analysis,
                    'annual_maintenance': pv_annual_maintenance,
                    'lifecycle_cost': sum(pv_cash_flows),
                    'npv': pv_npv,
                    'irr': pv_irr,
                    'lcoe': pv_lcoe,
                    'energy_production': energy_production
                },
                'diesel_analysis': {
                    **diesel_analysis,
                    'lifecycle_cost': sum(diesel_cash_flows),
                    'npv': diesel_npv,
                    'irr': diesel_irr,
                    'lcoe': diesel_lcoe
                },
                'comparison_metrics': {
                    'npv_difference': pv_npv - diesel_npv,
                    'irr_difference': pv_irr - diesel_irr,
                    'lcoe_difference': pv_lcoe - diesel_lcoe,
                    'payback_period': pv_analysis['initial_cost'] / max(diesel_annual_costs - pv_annual_maintenance, 1),
                    'cost_savings_20_years': diesel_npv - pv_npv
                },
                'environmental_impact': environmental_impact
            }
            
        except Exception as e:
            logger.error(f"Error in comprehensive analysis: {e}")
            raise

# Global service instance
techno_economic_service = TechnoEconomicService()

File: frontend/src/services/technoEconomicService.ts
interface TechnoEconomicParameters {
  // System sizing parameters
  systemSizingFactor: number;
  solarIrradianceFactor: number;
  ambientTemperature: number;
  windSpeed: number;
  includeSeasonalVariation: boolean;

  // Battery parameters
  batteryAutonomyFactor: number;
  batteryDepthOfDischarge: number;
  batteryType: 'lithium' | 'lead_acid';
  batteryRoundTripEfficiency: number;
  batteryLifetime: number;
  batteryTemperatureDerating: number;
  
  // System efficiency parameters
  inverterEfficiency: number;
  
  // Cost parameters
  panelCostPerWatt: number;
  panelCostPerKw: number;
  batteryCostPerKwh: number;
  inverterCostPerKw: number;
  structureCostPerKw: number;
  fixedCosts: number;
  
  // Economic parameters
  electricityTariff: number;
  fuelPriceEscalation: number;
  maintenanceCostEscalation: number;
  includeInflation: boolean;
  discountRate: number;
  projectLifetime: number;
}

interface CostingMethodology {
  method: 'perWatt' | 'fixedVariable' | 'componentBased';
  systemCostPerWatt?: number;
  panelCostPerWatt?: number;
  panelCostPerKw?: number;
  batteryCostPerKwh?: number;
  inverterCostPerKw?: number;
  structureCostPerKw?: number;
  fixedCosts?: number;
  numPanels?: number;
  panelRating?: number;
}

interface LoadProfile {
  dailyUsage: number;
  peakHours: number;
  equipment: Array<{
    name: string;
    power: number;
    hours: number;
    quantity: number;
    frequency: string;
    timeOfUse: string;
    critical: boolean;
  }>;
}

interface TechnoEconomicResult {
  pv: {
    initialCost: number;
    annualMaintenance: number;
    lifecycleCost: number;
    npv: number;
    irr: number;
    systemSize: number;
    batteryCapacity: number;
    costingMethod: string;
    panelCostPerWatt?: number;
    panelCostPerKw?: number;
    batteryCostPerKwh?: number;
    inverterCostPerKw?: number;
    structureCostPerKw?: number;
    fixedCosts?: number;
    numPanels?: number;
    panelRating?: number;
  };
  diesel: {
    initialCost: number;
    annualMaintenance: number;
    lifecycleCost: number;
    npv: number;
    irr: number;
  };
}

class TechnoEconomicService {
  private static readonly API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:3001/api';

  // Default parameters - can be overridden by user configuration
  private static readonly DEFAULT_PARAMETERS: TechnoEconomicParameters = {
    // System sizing parameters
    systemSizingFactor: 1.2,
    solarIrradianceFactor: 0.85,
    ambientTemperature: 25,
    windSpeed: 2.5,
    includeSeasonalVariation: true,
    
    // Battery parameters
    batteryAutonomyFactor: 1.0,
    batteryDepthOfDischarge: 0.8,
    batteryType: 'lithium',
    batteryRoundTripEfficiency: 0.92,
    batteryLifetime: 10,
    batteryTemperatureDerating: 0.05,
    
    // System efficiency parameters
    inverterEfficiency: 0.94,
    
    // Cost parameters
    panelCostPerWatt: 0.4,
    panelCostPerKw: 400,
    batteryCostPerKwh: 300,
    inverterCostPerKw: 300,
    structureCostPerKw: 150,
    fixedCosts: 0,
    
    // Economic parameters
    electricityTariff: 0.25,
    fuelPriceEscalation: 0.03,
    maintenanceCostEscalation: 0.025,
    includeInflation: true,
    discountRate: 0.08,
    projectLifetime: 20
  };

  static async performAnalysis(
    facilityId: number,
    costingMethod: CostingMethodology,
    parameters: Partial<TechnoEconomicParameters> = {},
    token: string
  ): Promise<{ success: boolean; data?: TechnoEconomicResult; error?: string }> {
    try {
      const requestData = {
        stage: costingMethod.method === 'componentBased' ? 'tendering' : 'prefeasibility',
        costingMethod: costingMethod.method,
        ...this.DEFAULT_PARAMETERS,
        ...parameters,
        ...costingMethod
      };

      const response = await fetch(`${this.API_BASE_URL}/techno-economic/${facilityId}`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(requestData),
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      return { success: true, data };
    } catch (error) {
      console.error('Techno-economic analysis failed:', error);
      return { 
        success: false, 
        error: error instanceof Error ? error.message : 'Unknown error occurred' 
      };
    }
  }

  static calculateLoadProfileFromSurvey(surveyData: any): LoadProfile {
    const equipmentPowerMap: { [key: string]: number } = {
      'lab_incubator': 150,
      'dry_steriliser': 1500,
      'mobile_phone': 5,
      'autoclave': 2000,
      'centrifuge': 300,
      'microscope': 50,
      'refrigerator': 200,
      'ultrasound': 500,
      'monitor': 10,
      'nebulizer': 50,
      'light': 40,
      'lamp': 40,
      'fan': 75,
      'computer': 65,
      'laptop': 65,
      'printer': 30,
      'scanner': 25,
      'x-ray': 3000,
      'suction': 200,
      'pump': 100
    };

    const getEquipmentPower = (name: string): number => {
      const lowercaseName = name.toLowerCase();
      for (const [key, power] of Object.entries(equipmentPowerMap)) {
        if (lowercaseName.includes(key)) {
          return power;
        }
      }
      return 100; // Default power rating
    };

    const calculateWeeklyUsageFactor = (frequency: string): number => {
      const frequencyMap: { [key: string]: number } = {
        'Daily': 1.0,
        'Weekly': 1/7,
        'Monthly': 1/30,
        'Rarely': 1/30
      };
      return frequencyMap[frequency] || 1.0;
    };

    let totalDailyUsage = 0;
    const processedEquipment = (surveyData.equipment || []).map((item: any) => {
      const power = getEquipmentPower(item.name);
      const quantity = parseInt(item.quantity) || 1;
      const hours = parseInt(item.hoursPerDay) || 8;
      const frequency = item.frequency || 'Daily';
      const timeOfUse = item.timeOfUse || 'day';
      
      const weeklyUsageFactor = calculateWeeklyUsageFactor(frequency);
      const dailyEnergy = (power * hours * quantity * weeklyUsageFactor) / 1000; // kWh
      
      totalDailyUsage += dailyEnergy;

      return {
        name: item.name,
        power,
        hours,
        quantity,
        frequency,
        timeOfUse,
        critical: item.critical || false
      };
    });

    // Calculate peak hours based on operational schedule
    const operationalHours = surveyData.operationalHours || { day: 8, night: 0 };
    const peakHours = Math.max(operationalHours.day, operationalHours.night) * 0.85;

    return {
      dailyUsage: totalDailyUsage,
      peakHours: Math.max(peakHours, 4), // Minimum 4 hours
      equipment: processedEquipment
    };
  }

  static getMockLoadProfile(): LoadProfile {
    return {
      dailyUsage: 15.5,
      peakHours: 6,
      equipment: [
        { name: 'LED Lights', power: 40, hours: 12, quantity: 10, frequency: 'Daily', timeOfUse: 'day', critical: true },
        { name: 'Refrigerator', power: 200, hours: 24, quantity: 1, frequency: 'Daily', timeOfUse: 'day', critical: true },
        { name: 'Computer', power: 65, hours: 8, quantity: 3, frequency: 'Daily', timeOfUse: 'day', critical: false },
        { name: 'Fans', power: 75, hours: 10, quantity: 4, frequency: 'Daily', timeOfUse: 'day', critical: false }
      ]
    };
  }

  static getMockAnalysisResult(): TechnoEconomicResult {
    return {
      pv: {
        initialCost: 8500,
        annualMaintenance: 170,
        lifecycleCost: 12100,
        npv: 11500,
        irr: 0.12,
        systemSize: 2.8,
        batteryCapacity: 19.4,
        costingMethod: 'perWatt',
        panelCostPerWatt: 0.4,
        panelCostPerKw: 400,
        batteryCostPerKwh: 300,
        inverterCostPerKw: 300,
        structureCostPerKw: 150,
        fixedCosts: 0
      },
      diesel: {
        initialCost: 3500,
        annualMaintenance: 2800,
        lifecycleCost: 59500,
        npv: 47200,
        irr: 0.05
      }
    };
  }

  static getDefaultParameters(): TechnoEconomicParameters {
    return { ...this.DEFAULT_PARAMETERS };
  }

  static getUserParameters(): TechnoEconomicParameters {
    return this.getMergedParameters();
  }

  static saveUserParameters(parameters: Partial<TechnoEconomicParameters>): void {
    localStorage.setItem('technoEconomicParams', JSON.stringify(parameters));
  }

  static loadUserParameters(): Partial<TechnoEconomicParameters> {
    const saved = localStorage.getItem('technoEconomicParams');
    return saved ? JSON.parse(saved) : {};
  }

  static getMergedParameters(): TechnoEconomicParameters {
    const userParams = this.loadUserParameters();
    return { ...this.DEFAULT_PARAMETERS, ...userParams };
  }

  static async performAnalysisFromLoadProfile(
    loadProfileData: any,
    methodology: string = 'perWatt',
    parameters: Partial<TechnoEconomicParameters> = {}
  ): Promise<TechnoEconomicResult> {
    try {
      // Convert load profile data to analysis parameters
      const systemSize = loadProfileData.peakDemand * 1.3; // 130% of peak demand
      const batteryCapacity = loadProfileData.dailyConsumption * 1.2; // 120% of daily consumption
      
      const mergedParams = { ...this.DEFAULT_PARAMETERS, ...parameters };
      
      const pvInitialCost = methodology === 'perWatt' 
        ? systemSize * 1000 * (mergedParams.panelCostPerWatt || 0.4)
        : (systemSize * (mergedParams.panelCostPerKw || 400)) + 
          (batteryCapacity * (mergedParams.batteryCostPerKwh || 300)) +
          (systemSize * (mergedParams.inverterCostPerKw || 300)) +
          (systemSize * (mergedParams.structureCostPerKw || 150)) +
          (mergedParams.fixedCosts || 0);
      
      const pvAnnualMaintenance = systemSize * 50; // $50/kW/year
      const pvLifecycleCost = pvInitialCost + (pvAnnualMaintenance * mergedParams.projectLifetime);
      
      // Diesel comparison
      const dieselInitialCost = loadProfileData.peakDemand * 800; // $800/kW for diesel generator
      const dieselAnnualMaintenance = loadProfileData.annualConsumption * 0.25; // $0.25/kWh for diesel + maintenance
      const dieselLifecycleCost = dieselInitialCost + (dieselAnnualMaintenance * mergedParams.projectLifetime);
      
      // Calculate NPV and IRR
      const annualSavings = dieselAnnualMaintenance - pvAnnualMaintenance;
      const pvNpv = this.calculateNPV(pvInitialCost, annualSavings, mergedParams.discountRate, mergedParams.projectLifetime);
      const pvIrr = this.calculateIRR(pvInitialCost, annualSavings, mergedParams.projectLifetime);
      
      return {
        pv: {
          initialCost: Math.round(pvInitialCost),
          annualMaintenance: Math.round(pvAnnualMaintenance),
          lifecycleCost: Math.round(pvLifecycleCost),
          npv: Math.round(pvNpv),
          irr: Math.round(pvIrr * 10000) / 10000,
          systemSize: Math.round(systemSize * 100) / 100,
          batteryCapacity: Math.round(batteryCapacity * 100) / 100,
          costingMethod: methodology,
          panelCostPerWatt: mergedParams.panelCostPerWatt,
          panelCostPerKw: mergedParams.panelCostPerKw,
          batteryCostPerKwh: mergedParams.batteryCostPerKwh,
          inverterCostPerKw: mergedParams.inverterCostPerKw,
          structureCostPerKw: mergedParams.structureCostPerKw,
          fixedCosts: mergedParams.fixedCosts
        },
        diesel: {
          initialCost: Math.round(dieselInitialCost),
          annualMaintenance: Math.round(dieselAnnualMaintenance),
          lifecycleCost: Math.round(dieselLifecycleCost),
          npv: Math.round(-pvNpv), // Negative of PV NPV
          irr: -0.02 // Negative IRR for diesel
        }
      };
    } catch (error) {
      console.error('Load profile analysis failed:', error);
      throw error;
    }
  }

  private static calculateNPV(initialCost: number, annualCashFlow: number, discountRate: number, years: number): number {
    let npv = -initialCost;
    for (let year = 1; year <= years; year++) {
      npv += annualCashFlow / Math.pow(1 + discountRate, year);
    }
    return npv;
  }

  private static calculateIRR(initialCost: number, annualCashFlow: number, years: number): number {
    // Simple IRR approximation using Newton-Raphson method
    let irr = 0.1; // Initial guess
    for (let i = 0; i < 100; i++) {
      let npv = -initialCost;
      let derivative = 0;
      for (let year = 1; year <= years; year++) {
        const discountFactor = Math.pow(1 + irr, year);
        npv += annualCashFlow / discountFactor;
        derivative -= (year * annualCashFlow) / (discountFactor * (1 + irr));
      }
      const newIrr = irr - npv / derivative;
      if (Math.abs(newIrr - irr) < 0.0001) {
        return newIrr;
      }
      irr = newIrr;
    }
    return irr;
  }
}

export default TechnoEconomicService;
export { TechnoEconomicService };
export type { TechnoEconomicParameters, CostingMethodology, LoadProfile, TechnoEconomicResult };
